### PR TITLE
Starboard 2.0

### PR DESCRIPTION
**Pull request by andrewj-brown**
Starboard 2.0, implementing a bunch of stuff I originally ceebs'ed, plus novel features that committee discovered a need for:
- Starboard reaction totals are now calculated as (reacts on original message + reacts on starboard message)
- These reacts are now enforced one-per-person *and* don't include the author of the original message
- Allowed starboarding messages from admin channels (comedy inbound!)
- Subsequently, changed the message format so that channels display the name, rather than a link (which will be invisible for admin/private channels)
- Committee now have the ability to "blacklist" a message from being starboarded
- Committee now have a "cleanup" command that should remove any race-condition-caused duplicate messages
- Deletions of bot-posted-starboard-messages no longer cause the bot to shit itself
- Misc bugfixes (timezones; ratelimit in .env.example; replies to deleted messages; empty messages)

Many of these features are only minimally tested. Prior to merging I'll need to get a few other people together in the testing server and make sure it's all fully functional.

### CODE REVIEW

File: uqcsbot/starboard.py
Comment (JamesDearlove) [inline uqcsbot/starboard.py:83]: Just a comment, is this worth logging in the admin-alerts channel as well?

File: uqcsbot/text.py
Comment (JamesDearlove) [inline uqcsbot/text.py:29]: 🫡

**Review by JamesDearlove — COMMENTED** [review]
Not a full review, but here's some comments

**Review by JamesDearlove — COMMENTED** [review]
Not a full review, but here's some comments

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:119]: I'm propagating None's through as "no message found". I could switch to an exception handling system, like blacklisted messages or deleted references?

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:385]: i considered this, but it has to return in the error case, so i'd only be switching an except/return for an if not none/return. is that worth it?

File: uqcsbot/text.py
Comment (andrewj-brown) [reply uqcsbot/text.py:29]: 🫡

**Comment by andrewj-brown** [comment]
alright i've made a few changes, mostly commenting. notably:
- function-block'ed the thing jimmy mentioned
- logging of blacklist in #admin-alerts
- fixed timezone imports
- insta-delete self-reacts when the bot sees them

File: uqcsbot/starboard.py
Comment (JamesDearlove) [reply uqcsbot/starboard.py:139]: Nah, that's fine. I didn't look too heavily at how the function is used

File: uqcsbot/starboard.py
Comment (JamesDearlove) [inline uqcsbot/starboard.py:256]: It appears that .history() returns an async generator, so await won't work here.

```
TypeError: object async_generator can't be used in 'await' expression
```

File: uqcsbot/starboard.py
Comment (JamesDearlove) [inline uqcsbot/starboard.py:113]: This is missing interaction.user and needs to reference the message directly.

```suggestion
        await self._blacklist_log(message, interaction.user, blacklist=False)
```

**Review by JamesDearlove — CHANGES_REQUESTED** [review]
Some additional problems I've found

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:278]: "async iterators" are hard

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:132]: so true

**Comment by andrewj-brown** [comment]
you can really tell i've spent so much time testing this pre-PR

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py]: yes

**Comment by JamesDearlove** [comment]
After conversations last night, I'm gonna mark this as a draft so it's not accidentally merged. Feel free to close it and start afresh @andrewj-brown

**Comment by andrewj-brown** [comment]
> After conversations last night, I'm gonna mark this as a draft so it's not accidentally merged. Feel free to close it and start afresh @andrewj-brown
I'll leave it up; honestly it would have been a draft from the start if I knew anything about Github's UI

File: uqcsbot/starboard.py
Comment (49Indium) [inline uqcsbot/starboard.py:319]: Maybe also note this at the definition of the error.

File: uqcsbot/starboard.py
Comment (49Indium) [inline uqcsbot/starboard.py:409]: Is this still required with the new discord gallery view?

File: uqcsbot/starboard.py
Comment (49Indium) [inline uqcsbot/starboard.py:513]: Is a handler for deleting messages also needed, or is this a subclass of another handler?

File: uqcsbot/starboard.py
Comment (49Indium) [inline uqcsbot/starboard.py:41]: Maybe change this to `pin_threshold`

**Review by 49Indium — COMMENTED** [review]
LGTM, but I'll wait for another review as well (probably from @JamesDearlove ).
Some of the terminology (such as "big" and "whitelist"/"blacklist") was a little confusing, so a brief set of definitions at the top or slight name changes may be nice.

File: .env.example
Comment (andrewj-brown) [reply .env.example:12]: https://github.com/UQComputingSociety/uqcsbot-discord/wiki/Tokens-and-Environment-Variables#sb_base_threshold-sb_big_threshold-sb_ratelimit

...this page should probably be better documented.

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:18]: I feel the comments are descriptive enough

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:49]: Fixed, thanks

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:116]: Any use of the blacklist should trigger a deletion (line 183)... but this is the dedicated "cleanup from errors" command, so I'll add this.

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:178]: Good idea. Added

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:221]: Yeah, added another ephemeral response. Ty for picking these up!

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:207]: Not fussed either way, but I've spent so much time looking at it that I just know it all, so perspectives of reviewers (i.e. similar perspective to committee in 5 years time wondering wtf we were on about) are very valuable.

Ctrl-H my beloved

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:319]: TBQH in the interests of reducing alert fatigue etc etc I'm just going to add an ID check there.

We want to avoid IDs wherever possible, so the bot behaves consistently across different servers. But this isn't going to cause errors if we run the bot in another server, and it's the best way to check genuine equality of two messages.

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:409]: I still think it's worth reducing image-spam on the starboard anyway. How often does a message with >2 attachments get sent to the discord, and how often would that message then get starred?

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:41]: My intention when writing was that the "big threshold" may not always mean pinning - in other servers, starboards have tiers where different emojis are edited into the starboard message, and committee may want to manually edit the pins or whatever. 

I can change big -> pin globally if it's a significant benefit to the readability, but the thresholds have deliberately action-agnostic names.

File: .env.example
Comment (49Indium) [reply .env.example:12]: I did not see that page. Maybe add a link to it within the code?

File: uqcsbot/starboard.py
Comment (49Indium) [reply uqcsbot/starboard.py:18]: The comments are descriptive, but I'm slightly against swear words within our code. It's up to you.

File: uqcsbot/starboard.py
Comment (49Indium) [reply uqcsbot/starboard.py]: I meant showing that the `SomethingsFucked` error should have a note that it will be called each time on the main server. I can see someone trying to debug this and not knowing why we are getting a  `SomethingsFucked` error.

File: uqcsbot/starboard.py
Comment (49Indium) [reply uqcsbot/starboard.py:418]: Good point. Up to you

File: .env.example
Comment (andrewj-brown) [reply .env.example:12]: It'd be better to link it more prominently in the README, because it serves as documentation for all our env variables (and in general, should also have more detailed instructions on everything else that the bot needs)

File: uqcsbot/starboard.py
Comment (49Indium) [reply uqcsbot/starboard.py:41]: Now that you've stated it, I agree with the action agnostic names. With the documentation you linked above (which I didn't see when I reviewed), this looks fine as is.

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:18]: Happy to be overruled by vote, but I don't think there's any point trying to maintain a veneer of professionalism when... *gestures hands at literally everything else*.

File: uqcsbot/starboard.py
Comment (andrewj-brown) [reply uqcsbot/starboard.py:522]: I might just add a wiki page with details about the starboard (because the list of different behaviours it has is getting quite extensive), and make the wiki links more obvious in the readme as a "START HERE" type thing.

File: uqcsbot/utils/err_log_utils.py
Comment (49Indium) [inline uqcsbot/utils/err_log_utils.py:18]: Love this!

**Review by 49Indium — APPROVED** [review]
LGTM

**Review by JamesDearlove — APPROVED** [review]
lgtm. Comments made reading this much easier, nice work

**Review by JamesDearlove — COMMENTED** [review]
image

Nice

**Pull request merged**
